### DISCUSSION TITLE
Trying to understand if `brew shellenv` is doing the right thing with zsh site-functions

### DISCUSSION THREAD

**Discussion by alichaudry** [Homebrew/discussions#5683] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.4.1
ORIGIN: https://github.com/Homebrew/brew
HEAD: 70672606c6ac07a5e8f75abeda7b8736e8285dbe
Last commit: 4 days ago
Core tap JSON: 18 Oct 03:06 UTC
Core cask tap JSON: 18 Oct 03:06 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 3.3.5 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.5/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 16.0.0 build 1600
Git: 2.39.5 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.7.1 => /usr/bin/curl
macOS: 15.0.1-arm64
CLT: 16.0.0.0.1.1724870825
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

I was trying to understand why zsh autocompletion wasn't working for me when I noticed that, unless I'm missing something (which may very well be the case), the `shellenv` logic doesn't really work as intended. Specifically, I see that [this](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cmd/shellenv.sh#L84-L87) logic relies on `HOMEBREW_SHELL_NAME` being `zsh`. However, `HOMEBREW_SHELL_NAME` comes from [here](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cmd/shellenv.sh#L18) which for me is always resolving to `login`:

```sh
~ /bin/ps -p $PPID -c -o comm=
login
```

Since it's returning `login`, this wouldn't match `zsh`, and therefore never add `${HOMEBREW_PREFIX}/share/zsh/site-functions` to my `fpath`, right? I can confirm that the line isn't appending anything to my fpath:

```sh
~ print -l $fpath | grep "$HOMEBREW_PREFIX"
~
```

Or maybe I'm thinking about this all wrong and `HOMEBREW_SHELL_NAME` comes from [here](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cmd/shellenv.sh#L14), but if I'm reading this right this would always be null as my `~/.zprofile` reads `eval "$(/opt/homebrew/bin/brew shellenv)"`, and there's no args being passed to `shellenv`.

I'd appreciate any insight here.

**Comment by carlocab** [comment, marked as answer]
The output of `/bin/ps -p $PPID -c -o comm=` when run directly in your terminal is not the same as what `brew shellenv` sees (your terminal and `brew shellenv` have different parent processes).

Can you try applying this patch to `$(brew --repository)/Library/Homebrew/cmd/shellenv.sh`
```diff
diff --git a/Library/Homebrew/cmd/shellenv.sh b/Library/Homebrew/cmd/shellenv.sh
index 498510ac44..d729c75be6 100644
--- a/Library/Homebrew/cmd/shellenv.sh
+++ b/Library/Homebrew/cmd/shellenv.sh
@@ -17,6 +17,7 @@ homebrew-shellenv() {
   else
     HOMEBREW_SHELL_NAME="$(/bin/ps -p "${PPID}" -c -o comm=)"
   fi
+  echo "HOMEBREW_SHELL_NAME=${HOMEBREW_SHELL_NAME}"
 
   if [[ -n "${HOMEBREW_MACOS}" ]] &&
      [[ "${HOMEBREW_MACOS_VERSION_NUMERIC}" -ge "140000" ]] &&
```
and then tell me the output of `brew shellenv | grep SHELL`?

You can do
```sh
git -C "$(brew --repository)" restore Library/Homebrew/cmd/shellenv.sh
```
after to undo the changes.

**Reply by alichaudry** [reply]
Thank you @carlocab, this was insightful. It does seem like it outputs `HOMEBREW_SHELL_NAME=-zsh` when I do it this way, which, looking at the new `shellenv`, should trigger the conditional [here](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cmd/shellenv.sh#L84).

On a side-note, `brew shellenv` wasn't returning anything at all, presumably because it's loaded when `.zprofile` is loaded and doesn't output anything on subsequent runs; I had to comment out the `.zprofile` command (`#eval "$(/opt/homebrew/bin/brew shellenv)"`), open a new shell, and manually run `/opt/homebrew/bin/brew shellenv`to see the output:

```sh
HOMEBREW_SHELL_NAME=-zsh
export HOMEBREW_PREFIX="/opt/homebrew";
export HOMEBREW_CELLAR="/opt/homebrew/Cellar";
export HOMEBREW_REPOSITORY="/opt/homebrew";
PATH="..."; export PATH;  # obfuscated
[ -z "${MANPATH-}" ] || export MANPATH=":${MANPATH#:}";
export INFOPATH="/opt/homebrew/share/info:${INFOPATH:-}";
```

Based on this, seems like my issue was because of:

1. A misunderstanding on my part of how my interactive shell's parent process is `login` whereas for brew env it's `-zsh`, and
2. @ZhongRuoyu's prescient comment below about how brew and brew docs are currently out-of-sync, but that I'm also on the stable version of brew and not the developer one. I will reply below on his comment.

**Comment by ZhongRuoyu** [comment]
Your frustration may be from the slight non-synchronisation between documentation and implementation, as pointed out by https://github.com/Homebrew/brew/issues/18579. While the ["`brew` Shell Completion"](https://docs.brew.sh/Shell-Completion) docs have been updated recently (https://github.com/Homebrew/brew/pull/18524), its relevant `fpath` code changes have not been released yet.

This is an acknowledged issue, and since you're following the stable releases, one option for you would be to wait until the next stable release (which, following Homebrew's typical release cycle, should take place early next week).

Alternatively, you can get the latest updates (though not officially considered stable yet) by first switching on the developer mode with `brew developer on`, and then `brew update`. After that, open a new shell to check if your `fpath` is set as expected. You may switch off the developer mode with `brew developer off` after the next stable release.

**Reply by alichaudry** [reply]
Thank you @ZhongRuoyu, this was indeed the issue (along with my misunderstanding of the source `/bin/ps -p "${PPID}" -c -o comm=` returning `login` for me but `-zsh` for how Homebrew runs the command).

Thank you for pointing out the recent changes (despite the doc updates) aren't out in the wild till ~early next week.

**Reply by phaseOne** [reply]
Updated brew on my macOS install today and completion functionality has been restored! TY

In `~/.zprofile`:
```
eval "$(/opt/homebrew/bin/brew shellenv)"
```

In ` ~/.zshrc`:
```
autoload -Uz compinit
compinit
```

**Comment by alichaudry** [comment]
Both @carlocab and @ZhongRuoyu are correct, and helped me solve my problem! (I can only mark one answer as "the answer" so I just picked the first response which was Carlo's.) Thank you both!
